public/app/plugins/datasource/elasticsearch/datasource.ts

```diff
@@ -898,17 +899,22 @@ export class ElasticDatasource
     let expression = query.query ?? '';
     switch (action.type) {
       case 'ADD_FILTER': {
-        if (expression.length > 0) {
-          expression += ' AND ';
-        }
-        expression += `${action.options.key}:"${action.options.value}"`;
+        // This gives the user the ability to toggle a filter on and off.
+        expression = queryHasFilter(expression, action.options.key, action.options.value)
+          ? removeFilterFromQuery(expression, action.options.key, action.options.value)
+          : addFilterToQuery(expression, action.options.key, action.options.value);
         break;
       }
       case 'ADD_FILTER_OUT': {
-        if (expression.length > 0) {
-          expression += ' AND ';
+        /**
+         * If there is a filter with the same key and value, remove it.
+         * This prevents the user from seeing no changes in the query when they apply
+         * this filter.
+         */
+        if (queryHasFilter(expression, action.options.key, action.options.value)) {
+          expression = removeFilterFromQuery(expression, action.options.key, action.options.value);
         }
-        expression += `-${action.options.key}:"${action.options.value}"`;
+        expression = addFilterToQuery(expression, action.options.key, action.options.value, '-');
         break;
       }
     }
```

Comment:
I am not sure if we can easily solve this for ES without a language/lucene parser. E.g. this will break queries like `(hostname:"hostname4" OR hostname:"hostname3") AND (label:"val3" OR label:"val4")`.

I know that the previous implementation of this in ES wasn't great, but at least it did not end in "broken" queries. WDYT about skipping ES for now and come back to this once we have a lucene parser? also @gabor WDYT?

Reply:
I agree. I think we could feature flag it until we implement the parser, which will follow up this.